torch/testing/_asserts.py

```diff
@@ -102,6 +102,36 @@ def wrapper(actual: Tensor, expected: Tensor, **kwargs: Any) -> Optional[Excepti
     return wrapper
 
 
+def _check_sparse_indices_and_values_individually(
+    check_tensors: Callable[..., Optional[Exception]]
+) -> Callable[..., Optional[Exception]]:
+    """Decorates dense tensor check functions to handle indices and values of sparse tensors individually.
+
+    If the inputs are not sparse, this decorator is a no-op.
+
+    Args:
+        check_tensors (Callable[[Tensor, Tensor], Optional[Exception]]): Tensor check function for dense
+        tensors.
+    """
+    @functools.wraps(check_tensors)
+    def wrapper(actual: Tensor, expected: Tensor, **kwargs: Any) -> Optional[Exception]:
+        if not actual.is_sparse:
+            return check_tensors(actual, expected, **kwargs)
+
+        actual = actual.coalesce()
+        expected = expected.coalesce()
```

Comment:
I also wonder, should we always coalesce here? How about this:
```
if expected.is_coalesced():
    actual = actual.coalesce()
    exc = check_tensors(actual.indices(), expected.indices(), **kwargs)
    ...
else:
    exc = check_tensors(actual._indices(), expected._indices(), **kwargs)
    ...
```
that is, the coalesce state of `expected` will determine if `coalesce` is required or not.

Reply:
Another option would be to add a `check_coalesced: bool = True` kwarg which would check `is_coalesced()` and would call `coalesce` in case the check is disabled. Thoughts?

Reply:
Ideally, `torch.testing.assert_close` should work without extra arguments for both coalesced and uncoalesced arguments. Otherwise, I would predict that the default coalescing will lead to false-positive results for tests where uncoalesced state is important but developers forget to specify `check_coalesced=False`.

Btw, I struggle to understand the specification of `check_coalesced`. How about `always_coalesce=True` option to force coalesce on `assert_close` inputs? When `always_coalesce=False`, then the coalesce state of `expected` and `actual` should be matching (uncoalesced or coalesced) for the `assert_close` to succeed.

Reply:
`check_coalesced` would have the same logic as the other `check_*` kwargs that we already have. That means if  `check_coalesced is True` we require `actual.is_coalesced() == expected.is_coalesced()` to match and bail out before any value comparison happens if it doesn't match. If on the other hand `check_coalesced = False` we would simply call `.coalesced()` on both `actual` and `expected` before we get to the member comparison.

For an example of this look at the checks for the dtype

https://github.com/pytorch/pytorch/blob/afdfd2288ab839ece89e667e44777201536b2d82/torch/testing/_asserts.py#L157-L158

and 

https://github.com/pytorch/pytorch/blob/afdfd2288ab839ece89e667e44777201536b2d82/torch/testing/_asserts.py#L185-L188

---

> Otherwise, I would predict that the default coalescing will lead to false-positive results for tests where uncoalesced state is important but developers forget to specify `check_coalesced=False`.

No strong opinion here. We could also go with the strict default `check_coalesced = True`.

Reply:
ok, now `check_coalesced` makes sense to me (I had an opposite understanding of it).

If `check_coalesced==True` leads to `actual.is_coalesced() == expected.is_coalesced() check` then that's good. Perhaps use `check_is_coalesced` because "check coalesced" can be misinterpreted.

If `check_coalasced==False` leads to `coalesce()` calls then that sounds like a side-effect which is not ideal.

The strict default `check_coalesced = True` means that tests need explicitly coalesce `actual` and `expected`. So, we would not need the kw argument anymore anyway.

I would vote for a strict `assert_close` version that has no "check_coalesced" argument and that would fail when `is_coalesce` attributes are mismatching (and use (`_indices`/`_values` for uncoalesced tensor comparisons). That would be analogous to comparing, say, tensors with different device parameters: even when the values of tensors are the same, `assert_close` will fail when the tensors are on different devices:
```
>>> a = torch.ones(3)
>>> b = torch.ones(3, device='cuda')
>>> torch.testing.assert_close(a, b)
Traceback (most recent call last):
<snip>
AssertionError: The values for attribute 'device' do not match: cpu != cuda:0.
```

Reply:
@cpuhrsch,  @mruberry  suggested you as tie breaker here. @pearu and I agree that by default sparse COO tensors should have a matching `is_coalesced()` attribute, but we disagree whether the user should be able to deactivate this or not.

Reply:
@pearu if you pass `check_device = False`, your snippet works fine, since `assert_close` will move both tensors to the CPU before the comparison:

```python
a = torch.ones(3)
b = torch.ones(3, device='cuda')
torch.testing.assert_close(a, b, check_device=False)
```

My proposal for `check_coalesced` would behave the same.

Reply:
Notice that `check_coalesced` reads as "check if `a` and/or `b` are coaleased" which is not what is actually done: "check if `a` and `b` are both either coalesced or uncoalesced".

Reply:
True, didn't think of that. `check_is_coalesced` is indeed a better choice.

Reply:
After some offline discussion @pearu and I agreed that a `check_is_coalesced: bool = True` is a good way to go. @cpuhrsch feel free to chime in if you disagree, but I will implement it for now.